File: LRU Cache/LRU Cache.cpp
class LRUCache{
public:

    class Entry{
        
        public:
        
        int key;
        Entry* before;
        Entry* next;
        
        Entry(int k){
            key = k;
            before = NULL;
            next = NULL;
        }
        
        ~Entry(){
            
        }
    };

    // key -> entry
    std::unordered_map<int, Entry*> entryMap;
    // key -> value
    std::unordered_map<int, int> cache;
    // For Accessing LinkedList
    Entry* listHead;
    Entry* listEnd;
    // Capacity
    int cap;

    LRUCache(int capacity) {
        cap = capacity;
        listHead = new Entry(0);
        listEnd = new Entry(0);
        listHead->next = listEnd;
        listEnd->before = listHead;
    }
    
    int get(int key) {
        if(cache.find(key) == cache.end()){
            return -1;
        }
        else{
            Entry* updatedEntry = entryMap[key];
            updatedEntry->before->next = updatedEntry->next;
            updatedEntry->next->before = updatedEntry->before;
            updatedEntry->next = listHead->next;
            updatedEntry->before = listHead;
            listHead->next->before = updatedEntry;
            listHead->next = updatedEntry;
            return cache[key];
        }
    }
    
    void set(int key, int value) {
        if(cache.find(key) == cache.end()){
            if(cap == cache.size()){
                // Remove the LRU entry
                int removedKey = listEnd->before->key;
                listEnd->before->before->next = listEnd;
                listEnd->before = listEnd->before->before;
                cache.erase(removedKey);
                entryMap.erase(removedKey);
            }
            Entry* newEntry = new Entry(key);
            listHead->next->before = newEntry;
            newEntry->next = listHead->next;
            listHead->next = newEntry;
            newEntry->before = listHead;
            entryMap.insert(std::make_pair(key, newEntry));
            cache.insert(std::make_pair(key, value));
        }
        else{
            // Update entry
            cache[key] = value;
            Entry* updatedEntry = entryMap[key];
            updatedEntry->before->next = updatedEntry->next;
            updatedEntry->next->before = updatedEntry->before;
            updatedEntry->next = listHead->next;
            updatedEntry->before = listHead;
            listHead->next->before = updatedEntry;
            listHead->next = updatedEntry;
        }
    }
};

File: LRU Cache/LRU Cache.java
public class LRUCache {
    
    class LLNode{
        
        int val;
        LLNode next;
        LLNode before;
        
        LLNode(){
            val = 0;
            next = null;
            before = null;
        }
        
        LLNode(int input){
            val = input;
            next = null;
            before = null;
        }
        
        LLNode(int input, LLNode nextNode, LLNode be4Node){
            val = input;
            next = nextNode;
            before = be4Node;
        }
    }
    
    // Store key -> value
    Map<Integer, Integer> map;
    // Store key -> LLNode
    Map<Integer, LLNode> lru;
    int cap;
    LLNode fakeHead;
    LLNode fakeEnd;
    
    public LRUCache(int capacity) {
        cap = capacity;
        map = new HashMap<Integer, Integer>();
        lru = new HashMap<Integer, LLNode>();
        fakeHead = new LLNode();
        fakeEnd = new LLNode();
        fakeHead.next = fakeEnd;
        fakeEnd.before = fakeHead;
    }
    
    public int get(int key) {
        if(map.containsKey(key)){
            LLNode getNode = lru.get(key);
            // Update getNode's position in Linkedlist
            getNode.before.next = getNode.next;
            getNode.next.before = getNode.before;
            getNode.next = fakeHead.next;
            fakeHead.next.before = getNode;
            getNode.before = fakeHead;
            fakeHead.next = getNode;
            return map.get(key);
        }
        else{
            return -1;
        }
    }
    
    public void set(int key, int value) {
        if(!map.containsKey(key) && (map.size() == cap)){
            // The LRU node needs to be discarded
            LLNode removedNode = fakeEnd.before;
            int removedNodeKey = fakeEnd.before.val;
            // removedNode is removed from Linkedlist
            removedNode.before.next = removedNode.next;
            removedNode.next.before = removedNode.before;
            // removedNode is removed from HashMap
            lru.remove(removedNodeKey);
            map.remove(removedNodeKey);
        }
        
        // Create LRU map entry if it is absent
        if(!map.containsKey(key)){
            LLNode newNode = new LLNode(key, fakeHead.next, fakeHead);
            fakeHead.next.before = newNode;
            fakeHead.next = newNode;
            lru.put(key, newNode);
        }
        
        // Target entry is updated || created
        map.put(key, value);
        // Linkedlist is updated
        LLNode targetNode = lru.get(key);
        // Remove targetNode from Linkedlist
        targetNode.before.next = targetNode.next;
        targetNode.next.before = targetNode.before;
        // targetNode is reinserted after fakeHead
        targetNode.next = fakeHead.next;
        fakeHead.next.before = targetNode;
        targetNode.before = fakeHead;
        fakeHead.next = targetNode;
    }
}
